File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    right = max(times) * n 
    left = 1 
    
    while left <= right:
        mid = (right+left)//2
        count = 0 
        
        for time in times:
            count += (mid // time)
            if count > n:
                break 
        
        if count >= n:
            right = mid -1 
        else:
            left = mid +1 
        
    answer = left
        
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;
class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        
        Arrays.sort(times);
        
        long left = 1; 
        long right = (long) Arrays.stream(times).max().getAsInt() * n;
        
        while(left <= right){
            // 종료 조건, 시간동
            long tmp = 0; 
            long mid = (long) (left + right) /2 ;
            
            for(int time: times){
                tmp += mid/time;
                
                if(tmp > n){
                    break;
                }
            }
            if (tmp >= n){
                answer = mid;
                right = mid - 1;
            }
            else{
                left = mid + 1;
            }            
        }
        return answer;
    }
}